front/src/applications/operationalStudies/utils.ts

```diff
@@ -293,3 +296,79 @@ export const isTooFast = (
   }
   return false;
 };
+
+/**
+ * Adjusts a step's arrival time based on its own tolerance values and the given direction.
+ *
+ * @param payload - The step containing timing data.
+ * @param direction - 'upstream' or 'downstream'
+ * @returns The updated step with adjusted arrival time and tolerances.
+ */
+export const adjustPayloadByDirection = (
+  payload: PostTimetableByIdStdcmApiArg,
+  direction: 'upstream' | 'downstream'
+): PostTimetableByIdStdcmApiArg => ({
+  ...payload,
+  body: {
+    ...payload.body,
+    steps: payload.body.steps.map((step, index, steps) => {
+      if (index !== 0 && index !== steps.length - 1) return step;
+      if (!step.timing_data) return step;
+
+      const {
+        arrival_time,
+        arrival_time_tolerance_before: beforeTolerance,
+        arrival_time_tolerance_after: afterTolerance,
+      } = step.timing_data;
+
+      const adjustment = direction === 'downstream' ? -beforeTolerance : afterTolerance;
+      const newArrivalTime = new Date(isoDateToMs(arrival_time) + adjustment).toISOString();
+
+      return {
+        ...step,
+        timing_data: {
+          ...step.timing_data,
+          arrival_time: newArrivalTime,
+          arrival_time_tolerance_before: direction === 'downstream' ? beforeTolerance : 0,
+          arrival_time_tolerance_after: direction === 'upstream' ? afterTolerance : 0,
+        },
+      };
+    }),
+  },
+});
+
+/**
+ * Adjusts the pathSteps of an input step based on the specified direction.
+ * For "upstream", it sets the "before" tolerance to zero and keeps the "after" tolerance.
+ * For "downstream", it sets the "after" tolerance to zero and keeps the "before" tolerance.
+ *
+ * @param simulationInputs - The original simulation inputs containing the pathSteps.
+ * @param direction - The direction, either "upstream" or "downstream".
+ * @returns The updated step with adjusted tolerances and arrival.
+ */
+export const adjustInputByDirection = (
+  simulationInputs: StdcmSimulationInputs,
+  direction: 'upstream' | 'downstream'
+): StdcmSimulationInputs => {
+  const adjustedPathSteps = simulationInputs.pathSteps.map((step) => {
+    if (step.isVia || !step.arrival || !step.tolerances) return step;
+
+    const toleranceBeforeMs = step.tolerances.before.valueOf();
+    const adjustedArrival =
+      direction === 'upstream'
+        ? dayjs(step.arrival).add(toleranceBeforeMs, 'millisecond').toDate()
```

Comment:
Here you add the `toleranceBeforeMs` but you should use the `toleranceAfterMs` I think